src/FHIR/SMART/ClientAdminController.php

```diff
@@ -558,4 +843,273 @@ private function renderFooter()
+                ,'type' => 'textarea'
+                ,'enabled' => true
+        ];
+        if (!empty($request['token'])) {
+            $parts = $this->parseTokenIntoParts($request['token']);
+            $databaseRecord = $this->getDatabaseRecordForToken($parts['jti'], $parts['token_type']);
+            if (!empty($databaseRecord)) {
+                $parts['client_id'] = $databaseRecord['client_id'];
+                $parts['status'] = $databaseRecord['revoked'] != 0 ? 'revoked' : $parts['status'];
+            }
+
+            $queryParams = ['token' => $databaseRecord['id'], 'clientId' => $databaseRecord['client_id']];
+            if ($parts['token_type'] == 'refresh_token') {
+                $parts['revoke_link'] = $this->getActionUrl([self::TOKEN_TOOLS_ACTION, self::REVOKE_REFRESH_TOKEN], ['queryParams' => $queryParams]);
+            } else {
+                $parts['revoke_link'] = $this->getActionUrl([self::TOKEN_TOOLS_ACTION, self::REVOKE_ACCESS_TOKEN], ['queryParams' => $queryParams]);
+            }
+            $parts['user_link'] = $this->getActionUrl(['edit', $databaseRecord['client_id']], ['fragment' => $databaseRecord['user_id']]);
+        }
+
+        // now let's grab our parser and see what we can do with all of this.
+        $this->renderTokenToolsHeader($request);
+        if (empty($databaseRecord)) {
+            ?>
+        <div class="alert alert-info">
+            <?php echo xlt("JWT not found in system"); ?>
+        </div>
+            <?php
+        }
+        ?>
+        <form method="POST" action="<?php echo $actionUrl; ?>">
+        <?php
+            $this->renderTextarea("token", $textSetting);
+        ?>
+        <input type="submit" class="btn btn-sm btn-primary" value="<?php echo xla("Parse Token"); ?>" />
+
+        <?php if (!empty($databaseRecord)) { ?>
+        <hr />
+        <h3><?php echo xlt("Token details"); ?>
+            <?php if ($databaseRecord['revoked'] == 0) : ?>
+              <a href="<?php echo attr($parts['revoke_link']); ?>"
+                       class="btn btn-sm btn-primary float-right" onclick="top.restoreSession()"><?php echo xlt('Revoke Token'); ?></a>
+            <?php endif; ?>
+        </h3>
+        <ul>
+        <li><?php echo xlt("Token JTI(DB token value)"); ?>: <?php echo $parts['jti']; ?></li>
+        <li><?php echo xlt("Token DB Id"); ?>: <?php echo $databaseRecord['id']; ?></li>
+        <li><?php echo xlt("Token Status"); ?>: <?php echo $parts['status']; ?></li>
+        <li><?php echo xlt("Token Type"); ?>: <?php echo $parts['token_type']; ?></li>
+        <li><?php echo xlt("Token Expiration"); ?>: <?php echo $databaseRecord['expiry']; ?></li>
+        <li><?php echo xlt("User UUID"); ?>:
+        <a href="<?php echo attr($parts['user_link']); ?>">
+            <?php echo $databaseRecord['user_id']; ?>
+        </a>
+        </li>
+        </ul>
+        <pre><?php echo json_encode($parts, JSON_PRETTY_PRINT); ?></pre>
```

Comment:
should wrap above with a text()
(at least I think we should since this is ouput in html context)